File: 1462_courseScheduleIV/main.cpp
#include <vector>
#include <queue>

class Solution {
public:
    std::vector<bool> checkIfPrerequisite(int numCourses, std::vector<std::vector<int>>& prerequisites, std::vector<std::vector<int>>& queries) {
        std::vector<bool> res;
        std::vector<std::vector<int>> graph(numCourses); 
        std::vector<int> indegree(numCourses, 0);
        std::vector<std::vector<bool>> table(numCourses, std::vector<bool>(numCourses,false));
        for(int i = 0; i < prerequisites.size(); ++i)
        {
            graph[prerequisites[i][0]].push_back(prerequisites[i][1]);
            indegree[prerequisites[i][1]]++;
        }

        std::queue<int> q;
        for(int i = 0;i < numCourses; ++i)
            if(indegree[i] == 0) 
                q.push(i);
        
        while(!q.empty())
        {
            int cur=q.front();
            q.pop();
            for(auto &it: graph[cur])
            {
                table[cur][it] = true;
                for(int i = 0; i < numCourses; ++i)
                    // i is prerequisite of cur 
                    if(table[i][cur] == true) 
                        table[i][it] = true;

                indegree[it]--;
                if(indegree[it] == 0)
                    q.push(it);
            }
        }
        for(int i = 0; i < queries.size(); ++i) 
            res.push_back(table[queries[i][0]][queries[i][1]]);
        
        return res;
    }
};

File: 1462_courseScheduleIV/main.py
from typing import List
import collections

class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        adj = [set() for _ in range(numCourses)]
        indegrees = [0] * numCourses
        ancestors = [set() for _ in range(numCourses)]

        for i in range(len(prerequisites)):
            pre, next = prerequisites[i][0], prerequisites[i][1]
            adj[pre].add(next)
            indegrees[next] += 1
        
        myQueue = collections.deque()
        for a in range(len(indegrees)):
            if(indegrees[a] == 0):
                myQueue.append(a)

        while myQueue:
            curr = myQueue.pop()
            neighbors = adj[curr]
            for neighbor in neighbors:
                indegrees[neighbor] -= 1
                ancestors[neighbor].add(curr)
                ancestors[neighbor].update(ancestors[curr])
                if(indegrees[neighbor] == 0):
                    myQueue.append(neighbor)

        res = []
        for query in queries:
            all_prereqs = ancestors[query[1]]
            if(query[0] in all_prereqs):
                res.append(True)
            else:
                res.append(False)
        
        return res